### ISSUE TITLE
Support for datasets in authenticated S3-compatible private buckets

### ISSUE THREAD

**Issue by unidesigner** [google/neuroglancer#507]
I am looking into supporting authenticated S3-Compatible file protocol where one could specify `accessKey` and `secretKey` to view data in a private bucket.

I imagine to be able to specify a source like this:

`zarr://https://s3.us-west-004.amazonaws.com/bucketname/dataset?s3_access_key_id=accessKey&s3_secret_access_key=secretKey`

which would initialize an `S3Client` from the [aws-sdk/client-s3](https://www.npmjs.com/package/@aws-sdk/client-s3) SDK and make the appropriate request to get the `info` file and data chunks.

Where would I get started to implement this in Neuroglancer?

**Comment by jbms** [comment]
The place to add support would be here:
https://github.com/google/neuroglancer/blob/master/src/neuroglancer/util/special_protocol_request.ts

However, there are a few issues to consider:
- You will need to know the bucket region in order to generate a correct signature.  It is a bit tricky to figure it out, see e.g. the approach we take in tensorstore (https://github.com/google/tensorstore/blob/4e82af4392fe4939875d80f9fa5fd01c55beda64/tensorstore/kvstore/s3/s3_endpoint.cc#L211) but here there is the added challenge that it needs to work within the cross-origin request limitations of the browser.  Alternatively you could require that the bucket region be specified manually as part of the datasource URL, though that is a bit annoying.
- If you put the access key in the datasource URL itself, then sharing the Neuroglancer URL/state will also share the access key.  That may in some cases be desired, but you would need to be careful to use an access key with only the limited privileges that you intend to grant.  It would be very easy to accidentally share a more privileged access key.  For GCS, I implemented ngauth, which allows neuroglancer users to access private GCS buckets.  It requires that you run an ngauth server, which verifies a given user has access to a bucket, and then provides a time-limited restrict authentication token.  For S3, I think there may be a way to accomplish something similar, without even the need for a server, by using AWS Cognito combined with suitable S3 access policies (https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_s3_cognito-bucket.html).  I expect in many cases that would be preferable to embedding the access token directly in the URL.  Nonetheless, it could make sense to still support directly embedding the access token in the URL, and that is ultimately simpler.
- A final issue is that the access token would ideally come at the beginning of the URL, before the bucket name, rather than at the end, so that completion works.

**Comment by aaronkanzer** [comment]
Hi @jbms -- it seems that when I click https://github.com/google/neuroglancer/blob/master/src/neuroglancer/util/special_protocol_request.ts -- I get a 404 -- any chance you know if the code moved?

Also, nice to meet you 👋 I work over at MIT with @kabilar and others. We are hoping to leverage neuroglancer in at least the short-term for viewing private zarrs stored in S3

**Comment by jbms** [comment]
Updated URL is here: https://github.com/google/neuroglancer/blob/master/src/util/special_protocol_request.ts

The `src/neuroglancer` prefix was renamed to `src/`.

If you decide on the approach you will use for getting the AWS credentials in Neuroglancer I can offer more advice.

**Comment by aaronkanzer** [comment]
> Updated URL is here: https://github.com/google/neuroglancer/blob/master/src/util/special_protocol_request.ts
> 
> The `src/neuroglancer` prefix was renamed to `src/`.
> 
> If you decide on the approach you will use for getting the AWS credentials in Neuroglancer I can offer more advice.

Any advice regarding AWS creds<>auth would be great, as all our assets are all hosted via S3 -- thanks in advance

Also, just tagging a few others involved in the project here for visibility @ayendiki @balbasty @MikeSchutzman

**Comment by jbms** [comment]
The options that I've thought of are in my previous comment: https://github.com/google/neuroglancer/issues/507#issuecomment-1844796884

The simplest thing to implement would be to use the syntax:

s3+awskey:<ACCESS_KEY>:<SECRET_KEY>://bucket/path

Additionally you could support Amazon Cogito for credentials --- that would probably be preferable in most cases but would be a bit more complicated.

You can probably use this library to handle the actual requests to s3:

https://www.npmjs.com/package/@aws-sdk/client-s3

There is also this example of using Amazon Cogito:
https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/getting-started-browser.html

**Comment by aaronkanzer** [comment]
Thanks @jbms -- could you just clarify the `s3+awskey:<ACCESS_KEY>:<SECRET_KEY>://bucket/path` option? Unless I'm mistaken, I don't think I'd want to present these keys in plaintext -- let me know what you had in mind.

Will look further into Cognito.

@unidesigner did you ever arrive at a solution? I am looking to proof-of-concept something, but curious to understand how you approached this.

Thanks all in advance

**Comment by unidesigner** [comment]
Hi @aaronkanzer - unfortunately, I have not had time to prioritize work on this. Are you working on a proof-of-concept implementation?

I'd suggest before adding complexity with Cognito to support the scheme proposed by @jbms to provide access_key and secret_key in the URL. The secret_key would indeed be provided in plain text applying [Security by Obscurity](https://en.wikipedia.org/wiki/Security_through_obscurity).

**Comment by aaronkanzer** [comment]
> Hi @aaronkanzer - unfortunately, I have not had time to prioritize work on this. Are you working on a proof-of-concept implementation?
> 
> I'd suggest before adding complexity with Cognito to support the scheme proposed by @jbms to provide access_key and secret_key in the URL. The secret_key would indeed be provided in plain text applying [Security by Obscurity](https://en.wikipedia.org/wiki/Security_through_obscurity).

@unidesigner Yes, we are working on a proof-of-concept. After some research, we are implementing [presigned cookies via AWS CloudFront](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-cookies.html). 

We have a CloudFront Distribution that then sits in front of our S3 bucket, and serve `neuroglancer` directly
This is allowing us to fetch a handful of private chunks of data in an efficient manner (e.g. in our case, we are working heavily with `.ome.zarr`.).

Once we have a cleaned-up e2e solution, I'm happy to share some diagrams or example code -- would also be curious to get @jbms thoughts too, and see if we can extend support directly into `neuroglancer` eventually.

Cc @kabilar

**Comment by aaronkanzer** [comment]
@unidesigner Realizing I never followed up here -- did you come to an implementation here?

We've been using our CloudFront solution for quite some time now with success -- let me know, happy to transfer any knowledge if helpful.

Cc @kabilar

**Comment by aaronkanzer** [comment]
> Hi [@aaronkanzer](https://github.com/aaronkanzer) - no, not yet unfortunately, but it is still something I'd want to look into given time. I'd be interested in understanding your CloudFront solution! Can you post it here or write me an email [git@unidesign.ch](mailto:git@unidesign.ch). Thank you!

@unidesigner -- sorry for the delay here! I've sent you an email via akanzer@mit.edu to hopefully meet!

**Comment by d-v-b** [comment]
I'm also very interested in this! @aaronkanzer would you have time to email me at [davis.v.bennett@gmail.com](mailto:davis.v.bennett@gmail.com), or is there code / documentation I could look at?

**Comment by joshmoore** [comment]
I'm interested as well.

**Comment by kabilar** [comment]
Hi all, thanks for your interest.  The high-level overview is documented [here](https://github.com/lincbrain/linc-archive/blob/master/doc/design/linc_permissions.md).  Would also be happy to meet.  Here is a [scheduling link](https://calendar.app.google/nCJTU47onQFYfzVV6) for Aaron and I.

**Comment by bentaculum** [comment]
Same here, would be interested in this feature.

**Comment by unidesigner** [comment]
Hi Matt @MatthewAitken
Will your pull request https://github.com/google/neuroglancer/pull/757/ solve the feature request people have? Would it also support other non-AWS S3-compatible object stores?

**Comment by MatthewAitken** [comment]
Hi @unidesigner 

#757 does allow a user to access private S3 buckets by using AWS signed requests, however due to the way the signature is calculated, it's not as simple as simply adding the access keys to the url. For that reason I also don't think it would support other non-AWS endpoints, unless they used the exact same signing and authentication strategy.

**Comment by unidesigner** [comment]
Hi @MatthewAitken - thank you for your feedback. Given your work, it may be possible to adapt the code for other S3-compatible services with minor modifications in the signature calculation. If I read your code correctly, it requires usage of the Python API to provide the access tokens to Neuroglancer and cannot be used without Python, is this correct? Is there a simple example which demonstrates its usage with an AWS S3 private bucket?

I wonder if encoding the access keys (and additional information like region/endpoint URLs required for some schemes etc.) via the source URL directly in Neuroglancer (i.e. not using the Python bindings), may not be the best solution and a more flexible solution along the lines of your code with a secret store that can be referenced in the URL would be better. The secrets could be stored in the JSON state? Perhaps read it once and store in the browser's local storage.

**Comment by MatthewAitken** [comment]
@unidesigner you are correct that currently it can't be used without the python server. The plan is to implement an s3+ngauth kvstore type, which points to an auth server that can return accesskeys and temporary tokens, e.g. assume-role-with-web-identity + cognito, which would then use the aws_sig4 code to sign the request. This mirrors the gs and gs+ngauth pattern.

I haven't pursued the URL only approach (s3+awskey:<ACCESS_KEY>:<SECRET_KEY>://bucket/path) because my use cases revolve around temporary access keys (needing the session token as well). I think it's definitely possible but using just long-term access and secret key pairs is not best practice, and as JBM mentioned makes it super easy to accidentally share credentials. But if it's an important use case, I could maybe create another example PR after the ngauth method is implemented.  

A third option would be a new UI window to allow the user to input credentials and/or to allow secrets in json state, and keep them completely locally but out of the URL, i.e. in a Session Storage. This would also more easily support session tokens, and also would probably be easier to adapt to other credential types. But I'm not super familiar with the security risks of that approach.
